src/name/ip_address.rs

```diff
@@ -62,3 +292,255 @@ pub(super) fn presented_id_matches_constraint(
 
     Ok(true)
 }
+
+pub(crate) fn is_valid_ipv4_address(ip_address: untrusted::Input) -> bool {
```

Comment:
So as mentioned earlier, here we're verying input, but it would be better to parse: the output should be a custom type that encapsulates both the contents (which we'll otherwise have to reevaluate again, with the risk of the logic going out of sync between both implementations) and the invariants enforced here.

Reply:
Thanks! I am aware of this great article. Here, I do wonder if this principle applies as great as on the original article. The reason is because the risk of going out of sync on "syntax" as immutable as IPv4/IPv6 addresses is very low. However, I also think going with the "parse, don't validate" methodology is a better way in general.

I am happy to look into that. Do you think we could contribute this later on a separate PR?

Reply:
I think it should become part of this PR. It doesn't seem that hard? You already have parse functions and validate functions, returning the parsed result directly from the validate functions seems fairly straightforward.

Reply:
Addressed in https://github.com/rustls/webpki/pull/21 (https://github.com/rustls/webpki/pull/21/commits/f1b6d66ba1fa8ef11aaf274ec24229280dcb17d9)